auth/nodesession.go

```diff
@@ -0,0 +1,155 @@
+			if err != nil {
+				logger.Log(0, "error during message writing:", err.Error())
+			}
+			return
+		}
+		user, err := isUserIsAllowed(loginMessage.User, loginMessage.Network, false)
+		if err != nil {
+			err = conn.WriteMessage(messageType, []byte(fmt.Sprintf("%s lacks permission to join.", loginMessage.User)))
+			if err != nil {
+				logger.Log(0, "error during message writing:", err.Error())
+			}
+			return
+		}
+		accessToken, err := requestAccessKey(loginMessage.Network, 1, user.UserName)
+		if err != nil {
+			req.Pass = fmt.Sprintf("Error from the netmaker controller %s", err.Error())
+		} else {
+			req.Pass = fmt.Sprintf("AccessToken: %s", accessToken)
+		}
+		// Give the user the access token via Pass in the DB
+		if err = netcache.Set(stateStr, req); err != nil {
+			logger.Log(0, "machine failed to complete join on network,", loginMessage.Network, "-", err.Error())
+			return
+		}
+	} else { // handle SSO / OAuth
+		redirectUrl = fmt.Sprintf("https://%s/api/oauth/register/%s", servercfg.GetAPIConnString(), stateStr)
+		err = conn.WriteMessage(messageType, []byte(redirectUrl))
+		if err != nil {
+			logger.Log(0, "error during message writing:", err.Error())
+		}
+	}
+
+	go func() {
+		for {
+			cachedReq, err := netcache.Get(stateStr)
+			if err != nil {
+				if strings.Contains(err.Error(), "expired") {
+					logger.Log(0, "timeout occurred while waiting for SSO on network", loginMessage.Network)
+					timeout <- true
+					break
+				}
+				continue
+			} else if cachedReq.Pass != "" {
+				logger.Log(0, "node SSO process completed for user", cachedReq.User, "on network", loginMessage.Network)
+				answer <- cachedReq.Pass
+				break
+			}
+			time.Sleep(500) // try it 2 times per second to see if auth is completed
+		}
+	}()
+
+	select {
+	case result := <-answer:
+		// a read from req.answerCh has occurred
+		err = conn.WriteMessage(messageType, []byte(result))
+		if err != nil {
+			logger.Log(0, "Error during message writing:", err.Error())
+		}
+	case <-timeout:
+		logger.Log(0, "Authentication server time out for a node on network", loginMessage.Network)
+		// the read from req.answerCh has timed out
+		err = conn.WriteMessage(messageType, []byte("Authentication server time out"))
+		if err != nil {
+			logger.Log(0, "Error during message writing:", err.Error())
+		}
+	}
+	// The entry is not needed anymore, but we will let the producer to close it to avoid panic cases
+	if err = netcache.Del(stateStr); err != nil {
+		logger.Log(0, "failed to remove node SSO cache entry", err.Error())
+	}
+	// Cleanly close the connection by sending a close message and then
+	// waiting (with timeout) for the server to close the connection.
+	err = conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""))
+	if err != nil {
+		logger.Log(0, "write close:", err.Error())
+		return
+	}
+	time.After(time.Second)
+	close(answer)
+	close(timeout)
+}
```

Comment:
you can close the channels in the `defer func()` otherwise any error above before this line will return without closing the channels